patches/api/0453-Brigadier-based-command-API.patch

```diff
@@ -0,0 +1,1090 @@
++     * @param argumentName argument name
++     * @param context context
++     * @return signed message
++     * @throws CommandSyntaxException syntax exception
++     */
++    @NotNull CompletableFuture<SignedMessage> resolveSignedMessage(@NotNull String argumentName, @NotNull CommandContext<CommandSourceStack> context) throws CommandSyntaxException;
++
++}
+diff --git a/src/main/java/io/papermc/paper/command/brigadier/argument/SelectorArgumentResolver.java b/src/main/java/io/papermc/paper/command/brigadier/argument/SelectorArgumentResolver.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..1161a854e795aa70e857c5fa39434cb5766f8bd5
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/command/brigadier/argument/SelectorArgumentResolver.java
+@@ -0,0 +1,13 @@
++package io.papermc.paper.command.brigadier.argument;
++
++import org.jetbrains.annotations.ApiStatus;
++
++/**
++ * Represents a selector argument which requires
++ * the context of a command source stack to be fully resolved.
++ * @param <T> type
++ */
++@ApiStatus.Experimental
++@ApiStatus.NonExtendable
++public interface SelectorArgumentResolver<T> extends ArgumentResolver<T> {
++}
+diff --git a/src/main/java/io/papermc/paper/command/brigadier/argument/VanillaArgumentProvider.java b/src/main/java/io/papermc/paper/command/brigadier/argument/VanillaArgumentProvider.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..4ecd1573c1b8c5ed5569d31714313e34be4125c9
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/command/brigadier/argument/VanillaArgumentProvider.java
+@@ -0,0 +1,73 @@
++package io.papermc.paper.command.brigadier.argument;
++
++import com.destroystokyo.paper.profile.PlayerProfile;
++import com.google.common.collect.Range;
++import com.mojang.brigadier.arguments.ArgumentType;
++import io.papermc.paper.math.BlockPosition;
++import net.kyori.adventure.key.Key;
++import net.kyori.adventure.text.Component;
++import net.kyori.adventure.text.format.NamedTextColor;
++import org.bukkit.GameMode;
++import org.bukkit.HeightMap;
++import org.bukkit.NamespacedKey;
++import org.bukkit.World;
++import org.bukkit.block.BlockState;
++import org.bukkit.entity.Entity;
++import org.bukkit.entity.Player;
++import org.bukkit.inventory.ItemStack;
++import org.bukkit.scoreboard.DisplaySlot;
++import org.checkerframework.checker.nullness.qual.NonNull;
++import org.checkerframework.framework.qual.DefaultQualifier;
++import org.jetbrains.annotations.ApiStatus;
++
++import java.util.Collection;
++import java.util.List;
++import java.util.UUID;
++
++@ApiStatus.Internal
++@DefaultQualifier(NonNull.class)
++interface VanillaArgumentProvider {
++
++    ArgumentType<SelectorArgumentResolver<Entity>> entity();
++
++    ArgumentType<SelectorArgumentResolver<Player>> player();
++
++    ArgumentType<SelectorArgumentResolver<List<Entity>>> entities();
++
++    ArgumentType<SelectorArgumentResolver<List<Player>>> players();
```

Comment:
I think we can reduce the generic issues by making all of these have the same type, just a `SelectorArgumentResolver` essentially. That would have methods to getSingleEntity getSinglePlayer, getEntities, getPlayers. I don't think that significantly reduces the "niceness" of having these generics but I think getting rid of as many generics as possible is best.

Reply:
I agree